ucm2: Support acp3xalc5682m98 on Chromebook "zork"

akihikodaki opened the pull request:
Signed-off-by: Akihiko Odaki \<akihiko.odaki@gmail.com\>

perexg commented:
I am afraid, but those files are too much Chromium sound server related. Some values are not defined in https://github.com/alsa-project/alsa-lib/blob/master/include/use-case.h#L287 and there are some oldies (like fixes hw:XXXX names instead hw:${CardId}) etc. Also the files should be placed outside conf.d subtree (probably the AMD/acp3x-alc5682 tree or so).

Basically, Chromium based UCM files are not acceptable for the upstream.

akihikodaki commented:
5534b2668432630686d5623d3e465f8563695c68:
- Removed undefined values, and introduced `${CardId}`. Also, renamed devices according to the convention. There are no discrepancies with alsa-lib anymore if I read `ucm_confdoc.h` and `use-case.h` correctly and didn't miss things.
- The files are moved to `AMD/acp3xalc5682m98`. (I'm not aware of the formal convention of the placement so please tell me if there is some documentation.)

I found jack kcontrol is missing while reviewing the files and submitted a patch. I'll submit a follow-up pull request for alsa-ucm-conf once it became clear the change is going to be included in the upstream release.
https://lore.kernel.org/lkml/20220407141316.5099-1-akihiko.odaki@gmail.com/

akihikodaki commented:
8686db62df271482992e17a6d1b1dde40bd5e73e: Added `JackControl`s as kcontrols were added to Linux 5.19 with https://github.com/torvalds/linux/commit/19aed2d6cdb72a7c92909832b9480d9cadebeef9.

perexg commented:
>     * The files are moved to `AMD/acp3xalc5682m98`. (I'm not aware of the formal convention of the placement so please tell me if there is some documentation.)

There is no formal convention, it should be just readable for other developers to identify the hardware. I think that adding of one more level like "AMD/acp3x/alc5682" may work (not sure what "m98" means).

perexg commented on ucm2/AMD/acp3xalc5682m98/HiFi/1mic.conf:
```diff
@@ -0,0 +1,89 @@
+		cset "name='Stereo1 ADC MIXL ADC1 Switch' 1"
+		cset "name='Stereo1 ADC MIXR ADC1 Switch' 1"
+	]
+	DisableSequence [
```
Remove empty sequences.

perexg commented on ucm2/AMD/acp3xalc5682m98/HiFi/1mic.conf:
```diff
@@ -0,0 +1,89 @@
+		PlaybackPCM "hw:${CardId},1"
+	}
+	EnableSequence [
+		cdev "hw:${CardId}"
```
All those cdev lines are extra (the device is set by default in the ucm code for hw: UCM cards). Remove all cdev lines from all sequences.

perexg commented on ucm2/AMD/acp3xalc5682m98/HiFi/2mic.conf:
```diff
@@ -0,0 +1,108 @@
+SectionVerb {
```
It seems that there are minimal changes in the configs. It would be probably much better to have one HiFi.conf configuration with If blocks to cover differences for this type of hw to simplify the future maintenance.

akihikodaki commented:
c4b8a49ea1990b7e28844636955f52b4744256ef applies suggested changes.

brendangreenley commented:
I am having Ubuntu sound issues with the exact same soundcard in a Lenovo C13 Yoga Chromebook with similar AMD specs. Would love to help test this, though am a bit of a noob.  I could start by adding Google_Zork / Morphius / 1510014998 into the regex statements to see if one of these configs matches mine?

akihikodaki commented:
@brendangreenley It should just work without any modification. Make sure you run Linux 5.19+, or have https://github.com/torvalds/linux/commit/19aed2d6cdb72a7c92909832b9480d9cadebeef9 to get jack unplug/plug detection work.

For Lenovo ThinkPad C13 Yoga Chromebook, I have several other patches to fix graphics glitch and rotation detections:
https://github.com/akihikodaki/edk2/tree/aki
https://github.com/systemd/systemd/pull/23830
https://patchew.org/linux/20220823001630.3973-1-akihiko.odaki@gmail.com/

akihikodaki commented:
edd33ae5fcc90804f94aca27b1722c87e0938209: I removed `JackControl` again as I submitted yet another change on the kernel side: https://lore.kernel.org/alsa-devel/20220924095025.7778-2-akihiko.odaki@gmail.com/

perexg commented:
Merged in fb226c6380e4a0d51b75ec43e51a8f65f5d44414 with the simplification in f21de3dcce80eab4b7383fe8237ef1136f242423 (I hope that my change is fine - report problems, please).

a-hurst commented:
@akihikodaki Looks like your separate headphone/mic jack patch [finally made it into the kernel](https://lore.kernel.org/all/20230802175737.263412-1-alpernebiyasak@gmail.com/t/#m1bf06fee3fab8200ddf4f5692a9edb93f48be893)! Are you still planning on adding JackControl back to the UCM now that it's in? Also, thanks a ton for your work getting all this upstreamed!
